prosemirror/model/from_dom.py

```diff
@@ -0,0 +1,1072 @@
+	def read_styles(self, styles: List[str]) -> Tuple[List[Mark], List[Mark]] | None:
+		add: List[Mark] = Mark.none
+		remove: List[Mark] = Mark.none
+
+		for i, style in enumerate(styles):
+			after: ParseRule | None = None
+			while i < len(styles) - 1:
+				rule = self.parser.match_style(styles[i], styles[i + 1], self, after)
+				if not rule:
+					break
+				if rule.ignore:
+					return None
+				if rule.clear_mark is not None:
+					for m in (self.top.pending_marks + self.top.active_marks):
+						if rule.clear_mark(m):
+							remove = m.add_to_set(remove)
+				else:
+					add = self.parser.schema.marks[rule.mark].create(rule.attrs).add_to_set(add)
+
+				if rule.consuming is False:
+					after = rule
+				else:
+					break
+			i += 2
+
+		return add, remove
+
+	def add_element_by_rule(self, dom_: DOMNode, rule: ParseRule, continue_after: ParseRule | None = None) -> None:
+		sync: bool = False
+		mark: Mark | None = None
+		node_type: NodeType | None = None
+
+		if rule.node is not None:
+			node_type = self.parser.schema.nodes[rule.node]
+			if node_type and not node_type.is_leaf:
+				sync = self.enter(
+					node_type,
+					rule.attrs,
+					rule.preserve_whitespace
+				)
+			elif node_type and not self.insert_node(
+				node_type.create(
+					rule.attrs
+				)
+			):
+				self.leaf_fallback(dom_)
+		elif rule.mark is not None:
+			mark_type = self.parser.schema.marks[rule.mark]
+			mark = mark_type.create(rule.attrs)
+			if mark is not None:
+				self.add_pending_mark(mark)
+
+		start_in = self.top
+		if node_type and node_type.is_leaf:
+			self.find_inside(dom_)
+		elif continue_after is not None:
+			self.add_element(dom_, continue_after)
+		elif rule.get_content is not None:
+			self.find_inside(dom_)
+			rule.get_content(dom_, self.parser.schema).for_each(lambda node: self.insert_node(node))
+		else:
+			content_dom = dom_
+
+			if isinstance(rule.content_element, str):
+				content_dom = dom_.cssselect(rule.content_element)
+			elif callable(rule.content_element):
+				content_dom = rule.content_element(dom_)
+			elif rule.content_element is not None:
+				content_dom = rule.content_element
+
+			self.find_around(dom_, content_dom, True)
+			self.add_all(content_dom)
+
+		if sync and self.sync(start_in):
+			self.open -= 1
+
+		if mark is not None:
+			self.remove_pending_mark(mark, start_in)
+
+	def add_all(self, parent: DOMNode, start_index: int | None = None, end_index: int | None = None) -> None:
+		index = start_index if start_index is not None else 0
+
+		try:
+			dom_ = list(parent)[index]
+		except IndexError:
+			pass
+		else:
+			end = None if end_index is None else list(parent)[end_index]
+
+			while dom_ != end:
+				self.find_at_point(parent, index)
+				self.add_DOM(dom_)
+
+				dom_ = dom_.getnext()
+				index += 1
+
+		self.find_at_point(parent, index)
+
+	def find_place(self, node: Node) -> bool:
+		route: List[NodeType] | None = None
+		sync: NodeContext | None = None
+
+		depth = self.open
+		while depth >= 0:
```

Comment:
I think there's other places in the code where we've opted for this pattern just because it's closer to the original code, although I think it would be fine to be a bit more idiomatic in these situations

Reply:
Yeah to me as long as the counter isn't modified in the loop body we should use `for` and `range` since otherwise it's possible to accidentally not update the counter (like by using `continue`)

Reply:
There's some slight differences still, because loop iteration variables aren't defined if the loop doesn't execute, i.e.:

```
for i in range(0):
    pass
print(i)
```

This throws an `UnboundLocalError`, whereas this will work with the other pattern.

Reply:
Yeah in some cases it would be necessary to declare them above the loop too